src/operators/Aggregate.cpp

```diff
@@ -0,0 +1,515 @@
+            arrow::ArrayVector filter_vector;
+
+            if (prev_filter != nullptr) {
+                for (int j = 0; j < prev_filter->num_chunks(); j++) {
+
+                    // Note that Compare cannot operate on ChunkedArrays.
+                    status = arrow::compute::And(&function_context,
+                                                 prev_filter->chunk(j),
+                                                 next_filter->chunk(j),
+                                                 &temp_filter);
+                    evaluate_status(status, __FUNCTION__, __LINE__);
+
+                    filter_vector.push_back(temp_filter.make_array());
+                }
+
+                prev_filter = std::make_shared<arrow::ChunkedArray>
+                        (filter_vector);
+            }
+            else {
+                prev_filter = next_filter;
+            }
+        }
+
+        return prev_filter;
+    }
+
+    void Aggregate::insert_group(int *its) {
+
+        arrow::Status status;
+        for (int i=0; i<group_type->num_children(); i++) {
+            switch(group_type->child(i)->type()->id()) {
+                case arrow::Type::STRING: {
+                    auto one_group_builder = (arrow::StringBuilder*)
+                            (group_builder->child(i));
+                    auto one_unique_values =
+                            std::static_pointer_cast<arrow::StringArray>
+                                    (unique_values_[i]);
+                    status = one_group_builder->Append(one_unique_values->GetString
+                            (its[i]));
+                    evaluate_status(status, __FUNCTION__, __LINE__);
+                    break;
+                }
+                case arrow::Type::INT64: {
+                    auto one_group_builder = (arrow::Int64Builder *)
+                            (group_builder->child(i));
+                    auto one_unique_values =
+                            std::static_pointer_cast<arrow::Int64Array>
+                                    (unique_values_[i]);
+                    status = one_group_builder->Append
+                            (one_unique_values->Value(its[i]));
+                    evaluate_status(status, __FUNCTION__, __LINE__);
+                    break;
+                }
+                default: {
+                    std::cerr << "Cannot insert unsupported aggregate type: "
+                    + group_type->child(i)->type()->ToString() << std::endl;
+                }
+            }
+        }
+
+        status = group_builder->Append(true);
+        evaluate_status(status, __FUNCTION__, __LINE__);
+    }
+
+    void Aggregate::insert_group_aggregate(arrow::compute::Datum aggregate) {
+
+        arrow::Status status;
+        // Append each group aggregate to aggregate_builder.
+        switch (aggregate.type()->id()) {
+            // Returns a Datum of type INT64
+            case arrow::Type::INT64: {
```

Comment:
Again for a future PR. We will soon have a lot of loops like this where there is a pattern: 

```
cast SomeArrowType: 
       someCommonCode1
       someCommonCode2_with_a_component_that_depends_on_SomeArrowType
       ...
```

I think there is a cleaner design (for the next PR, but would love for @NicholasCorrado @johnchronis and @gaffneyk to huddle on a design) that is based on [replacing conditional with polymorphism](https://refactoring.com/catalog/replaceConditionalWithPolymorphism.html), or if that turns out to be too expensive due to function call overhead, we can use a template (as a glorified macro layer). 

Note - You may have heard me talk about the need for a type-based structure in other context and I had hacked a quick prototype (not usable at all here) in this [repo](https://github.com/UWQuickstep/gift).